### PR TITLE
Bug logging using Azure Table Storage

### PR DESCRIPTION
**Pull request by arvindsingh-cloud**
## Description
</br>
 Brief description of the requested change.

## Checklist
- [ ] I have read the instructions mentioned in [Contribute to Code](/Contributing.md).
- [ ] I have read and understood the criteria described under [submitting changes](/Contributing.md#submitting-changes). 
- [ ] The title of the PR clearly describes the intent of the PR.
- [ ] This PR does not introduce any breaking changes to the code.

### CODE REVIEW

File: src/AzSK.ADO/Framework/Abstracts/ADOSVTBase.ps1 [hunk 1]
```diff
@@ -457,9 +457,16 @@ class ADOSVTBase: SVTBase {
 		#perform bug logging after control scans for the current resource
 		if ($BugLogParameterValue) 
 		{
-			if (($ControlResults.ControlResults.VerificationResult -contains "Failed") -or ($ControlResults.ControlResults.VerificationResult -contains "Verify")) {
-				$this.BugLoggingPostEvaluation($ControlResults, $BugLogParameterValue)
+			#added check azuretable check here, if (azuretable enabled and scan mode is ca) then only allow bug logging.
+			#If not cseo then go to bug logging. 
+			#if cseo then go to only if azuretable enabled and scan mode is ca
+			if (($this.SubscriptionContext.SubscriptionName -ne [Constants]::CSEOOrg) -or ( ($this.SubscriptionContext.SubscriptionName -eq [Constants]::CSEOOrg) -and [Helpers]::CheckMember($this.ControlSettings.BugLogging, "UseAzureStorageAccount") -and ($this.ControlSettings.BugLogging.UseAzureStorageAccount -eq $true -and ([AzSKSettings]::GetInstance().GetScanSource() -eq "CA") ) ) )
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Abstracts/ADOSVTBase.ps1:463]: enter only if UseAzureStorageAccount is true and scan mode is CA
or enter if UseAzureStorageAccount is false

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 [hunk 2]
```diff
@@ -11,6 +11,12 @@ class AutoBugLog {
     hidden [string] $BugLogParameterValue;
     hidden [string] $BugDescriptionField;
     hidden [string] $ServiceIdPassedInCMD;
+
+    hidden [bool] $UseAzureStorageAccount = $false;
+    hidden [BugLogCheckerHelper] $BugLogCheckerObj;
+    hidden [string] $ScanSource;
+    #will set this false if using azure table storage and does not have access on storage or ot activated.
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:18]: remove if not reqd

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 [hunk 3]
```diff
@@ -24,6 +30,21 @@ class AutoBugLog {
             $this.IsBugLogCustomFlow = $this.ControlSettings.BugLogging.BugAssigneeAndPathCustomFlow;
             $this.ServiceIdPassedInCMD = $InvocationContext.BoundParameters["ServiceId"];
         }
+        $This.ScanSource = [AzSKSettings]::GetInstance().GetScanSource();
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:33]: $this

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 [hunk 4]
```diff
@@ -24,6 +30,21 @@ class AutoBugLog {
...
+        #TODO: 
+        #$this.ScanSource = "CA";
+        #Check if UseAzureStorageAccount prop found in org-policy then set the local variable. 
+        #If UseAzureStorageAccount is true then initialize the BugLogCheckerObj singleton class object.
+        if ([Helpers]::CheckMember($this.ControlSettings.BugLogging, "UseAzureStorageAccount", $null)) {
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:38]: $null?

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 [hunk 5]
```diff
@@ -77,64 +98,65 @@ class AutoBugLog {
...
-                
-                        #compute hash of control Id and resource Id	
-                        $hash = $this.GetHashedTag($control.ControlItem.Id, $control.ResourceContext.ResourceId)
-                        #check if a bug with the computed hash exists
-                        #Removed ProjectName param and direcly added [BugLogPathManager]::BugLoggingProject, previously holding in variable and passing in method
-                        $workItem = $this.GetWorkItemByHash($hash, [BugLogPathManager]::BugLoggingProject)
-                        if ($workItem[0].results.count -gt 0) {
-                            #a work item with the hash exists, find if it's state and reactivate if resolved bug
-                            $this.ManageActiveAndResolvedBugs($ProjectName, $control, $workItem, $AssignedTo)
+                    if ($control.ControlItem.ControlID -notin $this.ControlSettings.BugLogging.ExcludeControlsInBugLogging) {
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:101]: remove this. Let us do it as a separate PR

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 [hunk 6]
```diff
@@ -313,6 +335,7 @@ class AutoBugLog {
 
     hidden [string] GetSecuritySeverity([string] $ControlSeverity) {
         $Severity = ""
+        #Changed severity of Moderate/Medium/Low to "2 - Important" becase only important/high SecuritySeverity bugs goes in s360
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:338]: let us do in separate PR

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 [hunk 7]
```diff
@@ -340,11 +363,21 @@ class AutoBugLog {
...
-		
-        $state = ($workItem[0].results.values[0].fields | where { $_.name -eq "State" })
-        $id = ($workItem[0].results.values[0].fields | where { $_.name -eq "ID" }).value
-
+        
+        #TODO: 
+        #$this.ScanSource = "CA";
+        $state = "";
+        $id = "";
+        if ($this.UseAzureStorageAccount -and $this.ScanSource -eq "CA")
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:371]: comment

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 [hunk 8]
```diff
@@ -430,39 +463,53 @@ class AutoBugLog {
...
-        else {
-            #resolved bug needs to be reactivated, hence search for new/active/resolved bugs
-            $body = '{"searchText":"{0}","skipResults":0,"takeResults":2,"sortOptions":[],"summarizedHitCountsNeeded":true,"searchFilters":{"Projects":["{1}"],"Work Item Types":["Bug"],"States":["Active","New","Resolved"]},"filters":[],"includeSuggestions":false}' | ConvertFrom-Json
+    hidden [object] GetWorkItemByHash([string] $hash, [string] $ProjectName) 
+    {
+        #TODO: 
+        #$this.ScanSource = "CA";
+        if ($this.UseAzureStorageAccount -and $this.ScanSource -eq "CA") 
+        {
+            return $this.BugLogCheckerObj.GetWorkItemByHashAzureTable($hash, $ProjectName, $this.ControlSettings.BugLogging.ResolvedBugLogBehaviour);
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:472]: ResolvedBugLogBehaviour - do we need this?

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+    hidden static [BugLogCheckerHelper] $BugLogCheckerInstance;
+
+    hidden [bool] $UseAzureStorageAccount;
+    hidden [string] $OrganizationName;
+    hidden [string] $StorageAccount;
+
+    hidden [object] $StorageAccountCtx;
+    hidden [string] $StorageRG;
+    hidden [bool] $hasAccessOnStorage = $true;
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1:12]: make it false and enable it only if access is found

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 (hunk 2 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:18]: removed

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 (hunk 3 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:33]: changed

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 (hunk 4 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:38]: not needed, removed

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 (hunk 6 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:336]: removed from here

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 (hunk 7 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:369]: added

File: src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1 (hunk 8 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/AutoBugLog.ps1:470]: removed

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        #$this.StorageAccount = "adoscannersa201021102003";
+        #$this.StorageRG = "ADOScannerRG";
+
+        #get storage
+        if ($this.StorageRG -and $this.StorageAccount) {
+            $keys = Get-AzStorageAccountKey -ResourceGroupName $this.StorageRG -Name $this.StorageAccount
+            $StorageContext = New-AzStorageContext -StorageAccountName $this.StorageAccount -StorageAccountKey $keys[0].Value -Protocol Https
+    
+            #$storageAcc = Get-AzStorageAccount -ResourceGroupName $this.StorageRG -Name $this.StorageAccount
+            $this.StorageAccountctx = $StorageContext.Context;
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1:31]: Ctx

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+                        $azTableBugInfo[0].IsDeleted = "Y";
+                        $azTableBugInfo[0] | Update-AzTableRow -Table $cloudTable;
+                    }
+
+                    return $bugObj;
+                }
+            }
+        }
+        catch {
+            Write-Host "Not able to access storage account."
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1:86]: Could not access storage account

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+    hidden [object] GetWorkItemByHashAzureTable([string] $hash, [string] $projectName, [string] $reactiveOldBug) 
+    {
+        #get table filter by name
+        $tableName = $this.GetTableName($projectName);
+        $bugObj = @(@{});
+        $bugObj[0].results = @();
+
+        try {
+            #get storage table.
+            $cloudTable = (Get-AzStorageTable -Name $tableName -Context $this.StorageAccountCtx).CloudTable;
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1:52]: debug

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    hidden [bool] InsertBugInfoInTable([string] $hash, [string] $projectName, [string] $ADOBugId) 
+    {
+        try 
+        {
+           $tableName = $this.GetTableName($projectName);
+
+           #Get table filterd by name.
+           $storageTables = Get-AzStorageTable -Context $this.StorageAccountCtx | Select Name;
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1:99]: debug

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+        }
+        
+        return $bugObj;
+    }
+
+    hidden [bool] InsertBugInfoInTable([string] $hash, [string] $projectName, [string] $ADOBugId) 
+    {
+        try 
+        {
+           $tableName = $this.GetTableName($projectName);
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1:96]: remove $projectname param

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+        }
+        catch {
+            return $false;
+        } 
+    }
+
+    hidden [bool] GetTableEntityAndCloseBug([string] $hash) 
+    {    
+        #get table filter by name
+        $tableName = $this.GetTableName("");
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1:122]: remove $projectname param

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+            if (!$this.errorMsgDisplayed) {
+               Write-Host "Not able to access storage account. Could not close the bug." -Red  
+            }
+            return $false;
+        }
+        
+        return $true;
+    } 
+
+    hidden [string] GetTableName([string] $resourceNameToMakeTableName)
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1:155]: remove $projectname param

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+                            $row.IsDeleted = "Y";
+                            $row | Update-AzTableRow -Table $cloudTable;
+                        }
+                    } 
+                }
+            }
+        }
+        catch {
+            if (!$this.errorMsgDisplayed) {
+               Write-Host "Not able to access storage account. Could not close the bug." -Red
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1:147]: Could not update entry of closed bug in storage table.

File: src/AzSK.ADO/Framework/BugLog/AutoCloseBugManager.ps1 [hunk 18]
```diff
@@ -2,9 +2,26 @@ Set-StrictMode -Version Latest
...
     hidden [string] $OrganizationName;
     hidden [PSObject] $ControlSettings;
+    hidden [string] $ScanSource;
+    hidden [bool] $UseAzureStorageAccount = $false;
+    hidden [BugLogCheckerHelper] $BugLogCheckerObj;
+
     AutoCloseBugManager([string] $orgName) {
         $this.OrganizationName = $orgName;
         $this.ControlSettings = [ConfigurationManager]::LoadServerConfigFile("ControlSettings.json");
+        $This.ScanSource = [AzSKSettings]::GetInstance().GetScanSource();
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/BugLog/AutoCloseBugManager.ps1:12]: $this

Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/AutoCloseBugManager.ps1:12]: changed

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 (hunk 9 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1:12]: changed

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 (hunk 10 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1:31]: changed

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 (hunk 16 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1]: removed

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 (hunk 17 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1]: changed

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 (hunk 15 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1]: removed

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 (hunk 14 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1]: removed

File: src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1 (hunk 11 above)
Comment (arvindsingh-cloud) [reply src/AzSK.ADO/Framework/BugLog/BugLogCheckerHelper.ps1]: changed msg

**Review by abhaydaga — APPROVED** [review]
